Support Block Adapter metrics with adapter_stats_id

itaigilo opened the pull request:
Closes # 8672.

## Change Description

Currently, the Block Adapters that use `reportMetrics` only report with `operation` and `isErrStr` labels -
Adding another `adapter_stats_id` label to it, to allow better tracing.

arielshaqed commented:
Not sure what I'm meant to review here: PR says "stats", code says "metrics".  I _think_ we need only to rename the PR - if it's about stats then we might need to do rather different things.

itaigilo commented:
> Not sure what I'm meant to review here: PR says "stats", code says "metrics". I _think_ we need only to rename the PR - if it's about stats then we might need to do rather different things.

Well, let's have it written:
When *you* write `stats`, you mean the `pkg/stats` package?
And when you write `metrics`, you mean everything that's grafana/prometheus related?

If so, not the mix, since the `stats.go` files are handling prometheus stuff...

Having said that, I don't mind renaming this PR 😄

guy-har commented on pkg/block/azure/stats.go:
```diff
@@ -1,31 +1,15 @@
-	durationHistograms.WithLabelValues(operation, isErrStr).Observe(time.Since(start).Seconds())
-	if sizeBytes != nil {
-		requestSizeHistograms.WithLabelValues(operation, isErrStr).Observe(float64(*sizeBytes))
+func NewAzureStats(adapterStatsID *string) block.Histograms {
```
```suggestion
func NewStats(adapterStatsID *string) block.Histograms {
```

guy-har commented on pkg/block/factory/build.go:
```diff
@@ -132,7 +132,7 @@ func BuildGSClient(ctx context.Context, params params.GS) (*storage.Client, erro
 }
 
-func buildGSAdapter(ctx context.Context, params params.GS) (*gs.Adapter, error) {
+func buildGSAdapter(ctx context.Context, params params.GS, adapterStatsID *string) (*gs.Adapter, error) {
```
```suggestion
func buildAdapter(ctx context.Context, params params.GS, adapterStatsID *string) (*gs.Adapter, error) {
```

guy-har reviewed (changes requested):
Looks good, requesting changes mostly because of the label name, that I'm not sure what it means

guy-har commented on pkg/block/metrics.go:
```diff
@@ -134,3 +137,49 @@ func (m *MetricsAdapter) GetRegion(ctx context.Context, storageID, storageNamesp
+	if adapterStatsID != nil {
+		labelNames = append(labelNames, "adapter_stats_id")
+	}
+	var durationHistograms = promauto.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Name: name + "_operation_duration_seconds",
+			Help: "durations of outgoing " + name + " operations",
+		},
+		labelNames,
+	)
```
I believe there is an issue here, each histogram can be initiated one time per name, otherwise it will panic. This means that:
1. We can't initiate a block adapter twice!
2. Don't think this can be used for 2 different adapterStatesIDs

guy-har replied:
OK, I see there's a nil check to ensure initialization happens only once. However, I still believe we should prioritize registering the metric during package-level initialization rather than run time

guy-har reviewed (changes requested):
Requesting changes again but this one is critical.
We can't initiate the same adapter twice, which raises the question, how was this tested? Did we try this out for two Adapters with same name and different AdapterStatID?

itaigilo commented:
I've done a major refactoring here - 
Histograms are only created once per name (which is per blockstore type).
Added comments about the limitations there.

Also renamed `stats` to `metrics` for clarity.

Tested it with two (S3) blockstores, and it reports each with a different label.

@guy-har PTAL again -
If it's still unclear, let's talk about it F2F.

arielshaqed commented on pkg/block/metrics.go:
```diff
@@ -134,3 +137,59 @@ func (m *MetricsAdapter) GetRegion(ctx context.Context, storageID, storageNamesp
 	return m.adapter.RuntimeStats()
 }
+
+type Histograms struct {
```
Consider calling this a Histogram - after all that's literally the name Prometheus gives it.

arielshaqed commented on pkg/block/metrics.go:
```diff
@@ -134,3 +137,59 @@ func (m *MetricsAdapter) GetRegion(ctx context.Context, storageID, storageNamesp
+		prometheus.HistogramOpts{
+			Name:    name + "_operation_size_bytes",
+			Help:    "handled sizes of outgoing " + name + " operations",
+			Buckets: prometheus.ExponentialBuckets(1, 10, 10), //nolint: mnd
```
TIL these the buckets we have. That's fairly nasty... (No action in this PR, of course.)

arielshaqed reviewed (commented):
Thanks! Guy's point is obviously blocking here.

guy-har commented on pkg/block/factory/build.go:
```diff
@@ -132,7 +132,7 @@ func BuildGSClient(ctx context.Context, params params.GS) (*storage.Client, erro
 }
 
-func buildGSAdapter(ctx context.Context, params params.GS) (*gs.Adapter, error) {
+func buildGSAdapter(ctx context.Context, params params.GS, metricsID *string) (*gs.Adapter, error) {
```
What is metricsID, Isn't it the adapter ID?

guy-har reviewed (commented):
I prefer not to initiate metrics on New adapter. I would recommend:
- leave the metrics as they were
- add the adapter id label
- initiated the adapter with a property of adapter id 
- call `reportMetrics` with adapterID
I believe this will reduce complexity
